xds/src/main/java/io/grpc/xds/XdsDependencyManager.java

```diff
@@ -0,0 +1,597 @@
+        typeWatchers = new TypeWatchers<>(type);
+        resourceWatchers.put(type, typeWatchers);
+      }
+
+      typeWatchers.add(resourceName, watcher);
+      xdsClient.watchXdsResource(type, resourceName, watcher);
+    });
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T extends ResourceUpdate> void cancelWatcher(XdsWatcherBase<T> watcher) {
+    if (watcher == null) {
+      return;
+    }
+
+    XdsResourceType<T> type = watcher.type;
+    String resourceName = watcher.resourceName;
+
+    this.syncContext.execute(() -> {
+      TypeWatchers<T> typeWatchers = (TypeWatchers<T>)resourceWatchers.get(type);
+      if (typeWatchers == null) {
+        logger.log(DEBUG, "Trying to cancel watcher {0}, but type not watched", watcher);
+        return;
+      }
+
+      typeWatchers.watchers.remove(resourceName);
+      xdsClient.cancelXdsResourceWatch(type, resourceName, watcher);
+    });
+
+  }
+
+  public void shutdown() {
+    for (TypeWatchers<?> watchers : resourceWatchers.values()) {
+      shutdownWatchersForType(watchers);
+    }
+    resourceWatchers.clear();
+  }
+
+  private <T extends ResourceUpdate> void shutdownWatchersForType(TypeWatchers<T> watchers) {
+    for (Map.Entry<String, XdsWatcherBase<T>> watcherEntry : watchers.watchers.entrySet()) {
+      xdsClient.cancelXdsResourceWatch(watchers.resourceType, watcherEntry.getKey(),
+          watcherEntry.getValue());
+    }
+  }
+
+  private void releaseSubscription(ClusterSubscription subscription) {
+    checkNotNull(subscription, "subscription");
+    String clusterName = subscription.getClusterName();
+    Set<ClusterSubscription> subscriptions = clusterSubscriptions.get(clusterName);
+    if (subscriptions == null) {
+      logger.log(DEBUG, "Subscription already released for {0}", clusterName);
+      return;
+    }
+
+    subscriptions.remove(subscription);
+    if (subscriptions.isEmpty()) {
+      clusterSubscriptions.remove(clusterName);
+      XdsWatcherBase<?> cdsWatcher =
+          resourceWatchers.get(CLUSTER_RESOURCE).watchers.get(clusterName);
+      cancelClusterWatcherTree((CdsWatcher) cdsWatcher);
+
+      maybePublishConfig();
+    }
+  }
+
+  private void cancelClusterWatcherTree(CdsWatcher root) {
+    checkNotNull(root, "root");
+    cancelWatcher(root);
+
+    if (root.getData() == null || !root.getData().hasValue()) {
+      return;
+    }
+
+    XdsClusterResource.CdsUpdate cdsUpdate = root.getData().getValue();
+    switch (cdsUpdate.clusterType()) {
+      case EDS:
+        String edsServiceName = cdsUpdate.edsServiceName();
+        EdsWatcher edsWatcher =
+            (EdsWatcher) resourceWatchers.get(ENDPOINT_RESOURCE).watchers.get(edsServiceName);
+        cancelWatcher(edsWatcher);
+        break;
+      case AGGREGATE:
+        for (String cluster : cdsUpdate.prioritizedClusterNames()) {
+          CdsWatcher clusterWatcher =
+              (CdsWatcher) resourceWatchers.get(CLUSTER_RESOURCE).watchers.get(cluster);
```

Comment:
What if this cluster is used by another aggregate cluster or used directly by the route? I think that same sort of problem could exist for EDS. Just because it isn't any more by this cluster doesn't mean it isn't used any more.

Reply:
I thought that clusters couldn't be in multiple aggregates.  I started implementing a more complicated tracking of who was using a cluster, but believed that the situations you listed weren't allowed so kept it simple.

Reply:
I'm not aware of such a restriction. Look into the requirements here.

Reply:
Done.  Cluster watchers now have a list of parents and only stop the XdsClientWatch when the parent list is empty.